File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cs
public class Solution {
    public int Search(int[] nums, int target) {
        return BinarySearch(nums, 0, nums.Length-1, target);
    }
 
    public int BinarySearch(int[] nums, int left, int right, int target){
        if (left > right) 
            return -1;

        int mid = left + (right-left)/2;

        if (target == nums[mid])
            return mid;

        if (nums[left] <= nums[mid]){
            if (nums[left] <= target && target < nums[mid]){
              return BinarySearch(nums, left, mid-1, target);
            } else {
              return BinarySearch(nums, mid+1, right, target);
            }
        } else {
            if (nums[mid] < target && target <= nums[right]){
              return BinarySearch(nums, mid+1, right, target);
            } else {
              return BinarySearch(nums, left, mid-1, target);
            }
        }
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        return binarySearch(nums, 0, nums.length - 1, target);
    }

    private int binarySearch(int[] nums, int lo, int hi, int target) {
        if (lo > hi) {
            return -1;
        }

        int mid = (lo + hi) / 2;

        if (target == nums[mid]) {
            return mid;
        }
    
        if (nums[lo] <= nums[mid]) { // left half is sorted
            if(nums[lo] <= target && target < nums[mid]) {
                return binarySearch(nums, lo, mid-1, target);
            } else {
                return binarySearch(nums, mid+1, hi, target);
            }
        } else { // right half is sortedz
            if(nums[mid] < target && target <= nums[hi]) {
                return binarySearch(nums, mid+1, hi, target);
            } else {
                return binarySearch(nums, lo, mid-1, target);        
            }
        }
    }
}